### PR TITLE
Add matrix transpose section

### PR DESCRIPTION
**Pull request by hdelan**
Section lifted from sc22 branch. 

CMake was broken for open source DPC++. This fixes it. I will also check if it works for icpx.

### CODE REVIEW

File: Code_Exercises/Matrix_Transpose/README.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+Compile with
+
+For DPC++ (using the Intel DevCloud):
+```sh
+dpcpp -I../../Utilities/include source.cpp
+
+```
+For DPC++ (not Intel DevCloud):
+```sh
+clang++ -fsycl -fsycl-targets=<triple> -I../../Utilities/include source.cpp
```
Comment (DuncanMcBain) [inline Code_Exercises/Matrix_Transpose/README.md:26]: Maybe this should be `icpx` (maybe both?)

File: Code_Exercises/Matrix_Transpose/solution.cpp [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+|localId[0]  |        |<----->*    |
++------------+        +------------+
+                      localId[1]
+
+*/
+
+#include <iostream>
+#include <vector>
+
+#if __has_include(<sycl/sycl.hpp>)
```
Comment (DuncanMcBain) [inline Code_Exercises/Matrix_Transpose/solution.cpp:49]: We can probably just simplify this to sycl/sycl.hpp now

File: Lesson_Materials/Matrix_Transpose/index.html [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+				<section>
+					<div class="hbox" data-markdown>
+						#### Coalesced global memory
+					</div>
+					<div class="container" data-markdown>
+						* Reading from and writing to global memory is generally very expensive.
+						* It often involves copying data across an off-chip bus.
+						  * This means you generally want to avoid unnecessary accesses.
+						* Memory access operations is done in chunks.
+						  * This means accessing data this is physically close together in memory is more efficient.
```
Comment (DuncanMcBain) [inline Lesson_Materials/Matrix_Transpose/index.html:82]:
```suggestion
						* Memory access operations are done in chunks.
						  * This means accessing data that is physically closer together in memory is more efficient.
```

File: Lesson_Materials/Matrix_Transpose/index.html [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+				</section>
+				<!--Slide 10-->
+				<section>
+					<div class="hbox" data-markdown>
+						#### Row-major vs Column-major
+					</div>
+					<div class="container" data-markdown>
+						* Coalescing global memory access is particularly important when working in multiple dimensions.
+						* This is because when doing so you have to convert from a position in 2d space to a linear memory space.
+						* There are two ways to do this; generally referred to as row-major and column-major.
```
Comment (DuncanMcBain) [inline Lesson_Materials/Matrix_Transpose/index.html:147]:
```suggestion
						* There are two ways to do this, generally referred to as row-major and column-major.
```

File: Lesson_Materials/Matrix_Transpose/index.html [hunk 5]
```diff
@@ -0,0 +1,455 @@
...
+					</div>
+					<div class="container" data-markdown>
+						![SYCL](./local_memory.png "SYCL")
+					</div>
+					<div class="container" data-markdown>
+						* The solution is local memory.
+						* Local memory is generally on-chip and doesn't have a cache as it's managed manually so is much lower latency.
+						* Local memory is a smaller dedicated region of memory per work-group.
+						* Local memory can be used to cache, allowing us to read from global memory just once and then read from local memory instead, often referred to as a scratchpad.
+						* Local memory can be accessed in an uncoalesced fashion without much performance degradation.
```
Comment (DuncanMcBain) [inline Lesson_Materials/Matrix_Transpose/index.html:197]: Do we want to mention bank conflicts here or is that too much for this tutorial?

File: Lesson_Materials/Matrix_Transpose/index.html [hunk 6]
```diff
@@ -0,0 +1,455 @@
...
+				<section>
+					<div class="hbox" data-markdown>
+						#### Synchronization
+					</div>
+					<div class="container">
+						<div class="col" data-markdown>
+							![SYCL](./barrier_1.png "SYCL")
+						</div>
+						<div class="col" data-markdown>
+							* Remember that work-items are not guaranteed to all execute at the same time (in parallel).
```
Comment (DuncanMcBain) [inline Lesson_Materials/Matrix_Transpose/index.html:252]:
```suggestion
							* Remember that work-items are not guaranteed to execute all at the same time (in parallel).
```

File: Lesson_Materials/Matrix_Transpose/index.html [hunk 7]
```diff
@@ -0,0 +1,455 @@
...
+				<!--Slide 32-->
+				<section>
+					<div class="hbox" data-markdown>
+						## Questions
+					</div>
+				</section>
+				<!--Slide 33-->
+				<section>
+					<div class="container" data-markdown>
+						Code_Exercises/Exercise_08_Matrix_Transpose
```
Comment (DuncanMcBain) [inline Lesson_Materials/Matrix_Transpose/index.html:438]: We've removed the exercise numbers

**Review by DuncanMcBain — APPROVED** [review]
Just minor suggestions but this is good

File: Lesson_Materials/Matrix_Transpose/index.html (hunk 5 above)
Comment (hdelan) [reply Lesson_Materials/Matrix_Transpose/index.html:197]: I think that's a bit too advanced. This exercise is meant to be the bare minimum intro to local mem and coalesced global memory

File: Code_Exercises/Matrix_Transpose/solution.cpp (hunk 2 above)
Comment (hdelan) [reply Code_Exercises/Matrix_Transpose/solution.cpp:49]: Good catch thanks

File: Code_Exercises/Matrix_Transpose/README.md (hunk 1 above)
Comment (hdelan) [reply Code_Exercises/Matrix_Transpose/README.md:26]: Yep good catch thanks

File: Lesson_Materials/Matrix_Transpose/index.html (hunk 7 above)
Comment (hdelan) [reply Lesson_Materials/Matrix_Transpose/index.html]: Good catch thanks.

**Comment by hdelan** [comment]
Thanks @DuncanMcBain changes made. If you give thumbs up I will merge this

**Comment by DuncanMcBain** [comment]
Go for it! It should be approved

**Pull request merged**